src/CLI.hs

```diff
@@ -0,0 +1,215 @@
+{-# LANGUAGE OverloadedStrings #-}
+module CLI (
+  LogLevel(..),
+  Component(..),
+  Command(..),
+  Options(..),
+
+  run
+) where
+
+import Data.Text hiding (empty)
+import Options.Applicative
+
+--------------------------------------------------------------------------------
+                             -- Types --
+--------------------------------------------------------------------------------
+data LogLevel
+  = Silent
+  | Panic
+  | Exception
+  | Error
+  | Warning
+  | Info
+  | Debug
+  | Vomit
+    deriving Show
+
+data Component
+  = Module Text
+  | Path FilePath
+
+data Command
+  = Languages
+  | Modules
+  | Flags
+  | Browse
+  | Check
+  | Expand
+  | CDebug
+  | DebugComponent
+  | CInfo
+  | Type
+  | Split
+  | Sig
+  | Refine
+  | Auto
+  | Find
+  | Lint
+  | Root
+  | Doc
+  | DumpSymbols
+  | Boot
+  | LegacyInteractive
+  | Version
+    deriving Show
+
+data Options =
+    Options { verbosity     :: LogLevel
+            , outputLisp    :: Bool
+            , lineSeparator :: Text
+            , linePrefix    :: (Text,Text)
+            , ghcOption     :: [Text]
+            , mapFile       :: (FilePath,FilePath)
+            , withGhc       :: FilePath
+            , withGhcPkg    :: FilePath
+            , withCabal     :: FilePath
+            , withStack     :: FilePath
+            , cmd           :: Command
+            }
+    deriving Show
+
+--------------------------------------------------------------------------------
+                           -- Interface --
+--------------------------------------------------------------------------------
+run :: IO Options
+run =
+  execParser (info optionsParser $ fullDesc <> progDesc "ghc-mod")
+
+parseLogLevel :: Int -> LogLevel
+parseLogLevel n
+  | n <= 0    = Silent
+  | n == 1    = Panic
+  | n == 2    = Exception
+  | n == 3    = Error
+  | n == 4    = Warning
+  | n == 5    = Info
+  | n == 6    = Debug
+  | otherwise = Vomit
+
+optionsParser :: Parser Options
+optionsParser =
+  Options <$> (silent' <|> verbosity')
```

Comment:
I'd prefer if these didn't duplicate the option names twice, i.e. just put the defintions inline. Here's an example of how I did this last time I had to: https://github.com/DanielG/kvm-in-a-box/blob/master/src/Options.hs#L33

Now using a bunch of custom operators with custom fixities might be crazy and ugly, not sure. I just don't want so much name duplication.